File: Framework.D-2015/Framework.D-2015/Persistencia/Conexion.vb
Imports System.Xml
Imports System.Reflection
Imports Framework.D_2015.Serializadores

Namespace Persistencia

    Public Class Conexion

        Public Enum EstrategiasDbEnum
            SqlServer
        End Enum

        Private _estrategiaConexion As IProveedorBD

        Public Sub New(estrategiaElegida As EstrategiasDbEnum, servidor As String, catalogo As String, usuario As String, clave As String)
            Dim configuracion As New ConfiguracionDB
            configuracion.Catalogo = catalogo
            configuracion.Clave = clave
            configuracion.Servidor = servidor
            configuracion.TipoDeConexion = estrategiaElegida
            configuracion.Usuario = usuario

            GenerarConexion(configuracion)

        End Sub

        Public Sub New(estrategiaElegida As EstrategiasDbEnum, servidor As String, catalogo As String)
            Dim configuracion As New ConfiguracionDB
            configuracion.Catalogo = catalogo
            configuracion.Servidor = servidor
            configuracion.TipoDeConexion = estrategiaElegida

            GenerarConexion(configuracion)

        End Sub

        Public Sub New(archivo As String)

            Dim archivoConf = System.IO.File.ReadAllText(archivo)

            Dim serializador As New SerializadorXml
            Dim configuracion As ConfiguracionDB = serializador.Deserealizar(Of ConfiguracionDB)(archivoConf)

            GenerarConexion(configuracion)

        End Sub

        Public Sub GenerarConexion(configuracion As ConfiguracionDB)
            Dim estrategiaElegida As EstrategiasDbEnum = configuracion.TipoDeConexion
            Select Case estrategiaElegida

                Case EstrategiasDbEnum.SqlServer
                    _estrategiaConexion = New ProveedorSql(configuracion.Servidor, configuracion.Catalogo)

            End Select
        End Sub

        Public Shared Sub GenerarArchivoConfiguracion(estrategiaElegida As EstrategiasDbEnum, servidor As String, catalogo As String, usuario As String, clave As String)
            Dim Configuracion As New ConfiguracionDB
            Configuracion.TipoDeConexion = estrategiaElegida
            Configuracion.Servidor = servidor
            Configuracion.Catalogo = catalogo
            Configuracion.Usuario = usuario
            Configuracion.Clave = clave

            Dim deserializador As New SerializadorXml
            deserializador.Serializar(Configuracion)
        End Sub

        Public Sub ConexionIniciar()
            _estrategiaConexion.ConexionIniciar()
        End Sub

        Public Sub ConexionFinalizar()
            _estrategiaConexion.ConexionFinalizar()
        End Sub

        Public Sub TransaccionIniciar()
            _estrategiaConexion.TransaccionIniciar()
        End Sub

        Public Sub TransaccionAceptar()
            _estrategiaConexion.TransaccionAceptar()
        End Sub

        Public Sub TransaccionCancelar()
            _estrategiaConexion.TransaccionCancelar()
        End Sub

        Public Sub EjecutarSinResultado(query As String, parametros As List(Of Parametro))
            _estrategiaConexion.EjecutarSinResultado(query, parametros)
        End Sub

        Public Function EjecutarEscalar(Of T)(query As String, parametros As List(Of Parametro)) As T
            Return _estrategiaConexion.EjecutarEscalar(Of T)(query, parametros)
        End Function

        Public Function EjecutarTupla(Of T As New)(query As String, parametros As List(Of Parametro)) As List(Of T)
            Return _estrategiaConexion.EjecutarTupla(Of T)(query, parametros)
        End Function

    End Class

End Namespace

File: Framework.D-2015/Framework.D-2015/Persistencia/Conexion.cs
using System.Collections.Generic;
using Framework.D_2015.Serializadores;
using Microsoft.VisualBasic.CompilerServices;

namespace Framework.D_2015.Persistencia
{
    public class Conexion
    {
        public enum EstrategiasDbEnum
        {
            SqlServer
        }

        private IProveedorBD _estrategiaConexion;

        public Conexion(EstrategiasDbEnum estrategiaElegida, string servidor, string catalogo, string usuario, string clave)
        {
            var configuracion = new ConfiguracionDB();
            configuracion.Catalogo = catalogo;
            configuracion.Clave = clave;
            configuracion.Servidor = servidor;
            configuracion.TipoDeConexion = ((int)estrategiaElegida).ToString();
            configuracion.Usuario = usuario;
            GenerarConexion(configuracion);
        }

        public Conexion(EstrategiasDbEnum estrategiaElegida, string servidor, string catalogo)
        {
            var configuracion = new ConfiguracionDB();
            configuracion.Catalogo = catalogo;
            configuracion.Servidor = servidor;
            configuracion.TipoDeConexion = ((int)estrategiaElegida).ToString();
            GenerarConexion(configuracion);
        }

        public Conexion(string archivo)
        {
            string archivoConf = System.IO.File.ReadAllText(archivo);
            var serializador = new SerializadorXml();
            var configuracion = serializador.Deserealizar<ConfiguracionDB>(archivoConf);
            GenerarConexion(configuracion);
        }

        public void GenerarConexion(ConfiguracionDB configuracion)
        {
            EstrategiasDbEnum estrategiaElegida = (EstrategiasDbEnum)Conversions.ToInteger(configuracion.TipoDeConexion);
            switch (estrategiaElegida)
            {
                case EstrategiasDbEnum.SqlServer:
                    {
                        _estrategiaConexion = new ProveedorSql(configuracion.Servidor, configuracion.Catalogo);
                        break;
                    }
            }
        }

        public static void GenerarArchivoConfiguracion(EstrategiasDbEnum estrategiaElegida, string servidor, string catalogo, string usuario, string clave)
        {
            var Configuracion = new ConfiguracionDB();
            Configuracion.TipoDeConexion = ((int)estrategiaElegida).ToString();
            Configuracion.Servidor = servidor;
            Configuracion.Catalogo = catalogo;
            Configuracion.Usuario = usuario;
            Configuracion.Clave = clave;
            var deserializador = new SerializadorXml();
            deserializador.Serializar(Configuracion);
        }

        public void ConexionIniciar()
        {
            _estrategiaConexion.ConexionIniciar();
        }

        public void ConexionFinalizar()
        {
            _estrategiaConexion.ConexionFinalizar();
        }

        public void TransaccionIniciar()
        {
            _estrategiaConexion.TransaccionIniciar();
        }

        public void TransaccionAceptar()
        {
            _estrategiaConexion.TransaccionAceptar();
        }

        public void TransaccionCancelar()
        {
            _estrategiaConexion.TransaccionCancelar();
        }

        public void EjecutarSinResultado(string query, List<Parametro> parametros)
        {
            _estrategiaConexion.EjecutarSinResultado(query, parametros);
        }

        public T EjecutarEscalar<T>(string query, List<Parametro> parametros)
        {
            return _estrategiaConexion.EjecutarEscalar<T>(query, parametros);
        }

        public List<T> EjecutarTupla<T>(string query, List<Parametro> parametros) where T : new()
        {
            return _estrategiaConexion.EjecutarTupla<T>(query, parametros);
        }
    }
}